Updates Next Framework Guide for deploying on Workers (#21256)
cloudflare/cloudflare-docs: src/content/docs/workers/frameworks/framework-guides/nextjs.mdx

---

import {
	Description,
	Details,
	Render,
	PackageManagers,
	Steps,
	WranglerConfig
} from "~/components";

> **Start from CLI** - scaffold an Next.js project on Workers.

Reviewer: looks like there's a next template that works with the deploy to cloudflare button?

https://github.com/cloudflare/templates/tree/staging/next-starter-template

(we really need to consolidate these templates...)

<PackageManagers
	type="create"
	pkg="cloudflare@latest my-next-app"
	args={"--framework=next --platform=workers"}
/>

:::note

> This is a simple getting started guide for a more detailed documentation on how the to use the Cloudflare Open Next adapter visit the [Open Next website](https://opennext.js.org/cloudflare).

Reviewer: ```suggestion
This is a simple getting started guide. For detailed documentation on how the to use the Cloudflare OpenNext adapter, visit the [OpenNext website](https://opennext.js.org/cloudflare).
```

:::

> ## What is Next.js

Reviewer: ```suggestion
## What is Next.js?
```

> [Next.js](https://nextjs.org/) is a [React](https://react.dev/) framework for building full stack applications. Think of React as the engine for building user interfaces (UIs) with components, while Next.js provides the chassis, structure, and additional features needed to build complete, production-ready web applications more easily and efficiently.

Reviewer: ```suggestion
[Next.js](https://nextjs.org/) is a [React](https://react.dev/) framework for building full stack applications.
```

(same feedback i had for the astro docs - this section should be as minimal as possible, because if you want to be persuaded to use Next.js, you'd go to their docs. If you're on this page, you probably know what Next is and have already decided you want to use it)

Next.js uses hybrid rendering to allow different parts of the your application to be rendered either on the server or on the client. For instance, critical pages needing fast initial loads and strong SEO might be server-rendered for faster load while highly interactive sections might leverage client-side rendering for dynamic updates without requiring full page reloads.

> Next supports Partial Prerendering to combine static and dynamic components in the same route.

Reviewer: Next.js supports Server-side and Client-side [rendering](https://nextjs.org/docs/app/building-your-application/rendering/server-components), as well as [Partial Prerendering](https://nextjs.org/docs/app/building-your-application/rendering/partial-prerendering) which lets you combine static and dynamic components in the same route. 

You can deploy your Next.js app to Cloudflare Workers using the OpenNext adaptor.

## Deploy a new Next.js project on Workers

<Steps>

1. **Create a new project with the create-cloudflare CLI (C3).**
	<PackageManagers
		type="create"
		pkg="cloudflare@latest my-next-app"
		args={"--framework=next --platform=workers"}
	/>

	<Details header="What's happening behind the scenes?">
		When you run this command, C3 creates a new project directory, initiates [Next.js's official setup tool](https://nextjs.org/docs/app/api-reference/cli/create-next-app), and configures the project for Cloudflare. It then offers the option to instantly deploy your application to Cloudflare.
	</Details>

2.  **Develop locally.**

	After creating your project, run the following command in your project directory to start a local development server.
	The command uses the Next.js development server. It offers the best developer experience by quickly reloading your app every time the source code is updated.

	<PackageManagers type="run" args="dev" />

> 3.  **Test your site with the Cloudflare adapter.**

Reviewer: ```suggestion
3.  **Test and preview your site with the Cloudflare adapter.**
```

> 	The command used in the previous step uses the Next.js development server to offer a great developer experience.
> 	However your application will run on Cloudflare Workers so you want to run your integration tests and verify that your application works correctly in this environment.

Reviewer: Maybe hide the detail in a dropdown?
```suggestion
<Detail title="What's the difference between dev and preview?">
	The command used in the previous step uses the Next.js development server, which runs in Node.js.

	However, your deployed application will run on Cloudflare Workers, which uses the `workerd` runtime. Therefore when running integration tests and previewing your application, you should use the preview command, which is more accurate to production, as it executes your application in the `workerd` runtime using `wrangler dev`. 
	</Detail>
```

	You can deploy your project to a [`*.workers.dev` subdomain](/workers/configuration/routing/workers-dev/) or a [custom domain](/workers/configuration/routing/custom-domains/) from your local machine or any CI/CD system (including [Workers Builds](/workers/ci-cd/#workers-builds)). Use the following command to build and deploy. If you're using a CI service, be sure to update your "deploy command" accordingly.

	<PackageManagers type="run" args="deploy" />

</Steps>

## Deploy an existing Next.js project on Workers

You can convert an existing Next.js application to run on Cloudflare

<Steps>

1. **Install [`@opennextjs/cloudflare`](https://www.npmjs.com/package/@opennextjs/cloudflare)**

	<PackageManagers type="install" pkg="@opennextjs/cloudflare@latest" />

2. **Add a Wrangler configuration file**

	In your project root, create a [Wrangler configuration file](/workers/wrangler/configuration/) with the following content:

	<WranglerConfig>
	```toml
	main = ".open-next/worker.js"
	name = "my-app"
	compatibility_date = "2025-03-25"
	compatibility_flags = ["nodejs_compat"]
	assets = { directory = ".open-next/assets", binding = "ASSETS" }
	```
	</WranglerConfig>

> 	:::note

Reviewer: Put this in a <Detail> component?

	As shown above, you must enable the [`nodejs_compat` compatibility flag](/workers/runtime-apis/nodejs/) _and_ set your [compatibility date](/workers/configuration/compatibility-dates/) to `2024-09-23` or later for your Next.js app to work with @opennextjs/cloudflare.
	:::

3. **Add a configuration file for OpenNext**

> 	In your project root, create an OpenNext configuration file name `open-next.config.ts` with the following content:

Reviewer: ```suggestion
	In your project root, create an OpenNext configuration file named `open-next.config.ts` with the following content:
```

---

import {
	Description,
	Details,
	Render,
	PackageManagers,
	Steps,
	WranglerConfig
} from "~/components";

> **Start from CLI** - scaffold an Next.js project on Workers.

Reviewer 2: ??
```suggestion
**Start from CLI** - scaffold a Next.js project on Workers.
```
